sqlx-core/src/connection.rs

```diff
@@ -1,30 +1,165 @@
+pub trait Connection
+where
+    Self: Send + 'static,
+{
+    type Database: Database;
+
     /// Close this database connection.
     fn close(self) -> BoxFuture<'static, crate::Result<()>>;
 
     /// Verifies a connection to the database is still alive.
-    fn ping(&mut self) -> BoxFuture<crate::Result<()>> {
-        Box::pin(self.execute("SELECT 1", Default::default()).map_ok(|_| ()))
-    }
+    fn ping(&mut self) -> BoxFuture<crate::Result<()>>;
+
+    #[doc(hidden)]
+    fn describe<'e, 'q: 'e>(
+        &'e mut self,
+        query: &'q str,
+    ) -> BoxFuture<'e, crate::Result<Describe<Self::Database>>>;
 }
 
 /// Represents a type that can directly establish a new connection.
-pub trait Connect {
-    type Connection: Connection;
-
+pub trait Connect: Connection {
     /// Establish a new database connection.
-    fn connect<T>(url: T) -> BoxFuture<'static, crate::Result<Self::Connection>>
+    fn connect<T>(url: T) -> BoxFuture<'static, crate::Result<Self>>
     where
         T: TryInto<Url, Error = crate::Error>,
         Self: Sized;
 }
+
+mod internal {
+    pub enum MaybeOwnedConnection<'c, C>
+    where
+        C: super::Connect,
+    {
+        Borrowed(&'c mut C),
+        Owned(super::PoolConnection<C>),
+    }
+
+    pub enum ConnectionSource<'c, C>
+    where
+        C: super::Connect,
+    {
+        Empty,
+        Connection(MaybeOwnedConnection<'c, C>),
+        Pool(super::Pool<C>),
+    }
+}
+
+pub(crate) use self::internal::{ConnectionSource, MaybeOwnedConnection};
+
+impl<'c, C> MaybeOwnedConnection<'c, C>
+where
+    C: Connect,
+{
+    pub(crate) fn borrow(&mut self) -> MaybeOwnedConnection<'_, C> {
+        match self {
+            MaybeOwnedConnection::Borrowed(conn) => MaybeOwnedConnection::Borrowed(&mut *conn),
+            MaybeOwnedConnection::Owned(ref mut conn) => MaybeOwnedConnection::Borrowed(conn),
+        }
+    }
+}
+
+impl<'c, C, DB> ConnectionSource<'c, C>
+where
+    C: Connect<Database = DB>,
+    DB: Database<Connection = C>,
+{
+    pub(crate) async fn resolve_by_ref(&mut self) -> crate::Result<MaybeOwnedConnection<'_, C>> {
+        if let ConnectionSource::Pool(pool) = self {
+            *self =
+                ConnectionSource::Connection(MaybeOwnedConnection::Owned(pool.acquire().await?));
+        }
+
+        Ok(match self {
+            ConnectionSource::Empty => panic!("`PgCursor` must not be used after being polled"),
+            ConnectionSource::Connection(conn) => conn.borrow(),
+            ConnectionSource::Pool(_) => unreachable!(),
+        })
+    }
+
+    pub(crate) async fn resolve(mut self) -> crate::Result<MaybeOwnedConnection<'c, C>> {
+        if let ConnectionSource::Pool(pool) = self {
+            self = ConnectionSource::Connection(MaybeOwnedConnection::Owned(pool.acquire().await?));
+        }
+
+        Ok(self.into_connection())
+    }
+
+    pub(crate) fn into_connection(self) -> MaybeOwnedConnection<'c, C> {
```

Comment:
Perhaps it would be good to document the expectations of calling this method. 2/3 variants of the match would lead to panic if `resolve` || `resolve_by_ref` are not called before hand (as far as I can tell at this point in the review). Looks like it is called once outside of this module as well. Thoughts?

Reply:
This has been cleaned up and can't panic now.